### DISCUSSION TITLE
Handle current session when auth via new provider

### DISCUSSION THREAD

**Discussion by lavcraft** [nextauthjs/next-auth#2976] [Help] [answered]
Hello, i can't find how to handle my case in documentation.
My case:
1. SignIn via Custom OAuth 2 Provider
2. SignIn via Google account after
I need to save access_token`s to current session, but after second signIn, new session is completely replace old.

How to access to current authentication data before create new session in callbacks?

**Comment by balazsorban44** [comment, marked as answer]
if you are using a stateless cookie persisted JWT session (which is the default), you don't have that option. you will need a database/adapter to persist the user data between sessions https://next-auth.js.org/adapters/overview

**Reply by lavcraft** [reply]
Sounds reasonable, but what do you think about this ability in stateless mode? In my opinion it can be handled too (may be in special callback)

**Reply by balazsorban44** [reply]
Here is the relevant core code:

https://github.com/nextauthjs/next-auth/blob/58a98b667dd1be3c801220b27f965025ac795d75/src/server/routes/callback.js#L88-L108

You can see on line 89 that we have the previous token in scope, but on line 96 we create a new `defaultToken` anyway.

I think it would make sense to re-use the previous in place of the `defaultToken`, and the change is extremely easy (`defaultToken = sessionToken ?? {...}`. This, but for all three flows, found in the same file.), but the docs should be updated accordingly. This might also be a breaking change as it changes the expectations about the `token` property.

Another way of doing this would be to juggle the names `token`, `prevToken`, `defaultToken`, `sessionToken` and find a combination that results in the least amount of change for current users, and the least confusion about the difference between the default (id, name, email, picture) token and the previous token, which potentially can contain very different things.

**Reply by lavcraft** [reply]
May be. But more flexible way - get ability to decide, which token/session is appropriate, to user. Because in some cases user want replace session, but in other cases want to merge or extend with previous. Also, this approach provide more safely backward compatibility

**Reply by balazsorban44** [reply]
Want to write up a feature request/RFC? We could ask the people what they think.

**Reply by lavcraft** [reply]
Sure, where is guide for feature request?

**Reply by balazsorban44** [reply]
https://github.com/nextauthjs/next-auth/issues/new?assignees=&labels=enhancement&template=feature_request.yaml

**Reply by lavcraft** [reply]
Done: https://github.com/nextauthjs/next-auth/issues/2988
